refine ut framework including Part 1 and Part 2

binmahone opened the pull request:
This PR is a superset of #10851, it closes #10850

Compared to #10851 . it has one more additional commit to distinguish two modes for expression based suites: vectorized parameter mode and scalar parameter mode. 

For RAPIDS expressions that only support scalar parameters (litOnlyTypes) , we have to use scalar parameter mode. For others we prefer to use vectorized parameter mode.

In case I can't get approval from reviewer because of the additional commit before 24.06 code freeze, we can merge #10851 , otherwise merge this PR.

binmahone commented:
build

winningsix commented on tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsTestsTrait.scala:
```diff
@@ -32,16 +33,18 @@ import org.apache.spark.sql.{Column, DataFrame, Row, SparkSession}
 import org.apache.spark.sql.types._
 import org.apache.spark.unsafe.types.UTF8String
 
+
```
nit: extra line?

binmahone replied:
fixed

winningsix commented on tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/RapidsSQLTestsBaseTrait.scala:
```diff
@@ -121,6 +121,12 @@ object RapidsSQLTestsBaseTrait {
       .set("spark.sql.cache.serializer", "com.nvidia.spark.ParquetCachedBatchSerializer")
+      .set("spark.sql.session.timeZone", "UTC")
+      .set("spark.rapids.sql.explain", "ALL")
+      // uncomment below config to run `strict mode`, where fallback to CPU is treated as fail
```
nit: do we want a parameterized test suit here for both `strict mode` and non-`strict mode` associated with different `RapidsTestSettings`?

binmahone replied:
A lot more test cases will fail in strict mode as shown in our report, and we'll need even more resource to take care of the failed test cases in strict mode. 

Now that we're still struggling with non-strict mode, the result of strict mode is more of a reference than a guide to our daily priority. So it's not mandatory to do so at present? we can add that anytime when we have enough resource to deal with failed test cases in strict mode .

binmahone replied:
Also we should consider how to parallel execute these two modes to avoid CI taking too much time

binmahone commented:
build

binmahone commented:
build

jlowe commented:
Having two PRs pending that do very similar things is confusing, especially when one of them already has a lot of active comments.  Why not post a separate PR that just has the last commit?  Yes, there's a small amount of conflicts to resolve when one goes in before the other, but fairly straightforward to resolve.  As it is now, this could get approved and merged without addressing all of the comments in #10851.

binmahone commented:
> Having two PRs pending that do very similar things is confusing, especially when one of them already has a lot of active comments. Why not post a separate PR that just has the last commit? Yes, there's a small amount of conflicts to resolve when one goes in before the other, but fairly straightforward to resolve. As it is now, this could get approved and merged without addressing all of the comments in #10851.

Hi Jason, I also considered your approach, but that would result in a PR against binmahone/spark-rapids instread of NVIDIA/spark-rapids ? But I have to admit these two PRs are indeed consufing.. I had to do this because of the deadline of code freeze.

"As it is now, this could get approved and merged without addressing all of the comments in https://github.com/NVIDIA/spark-rapids/pull/10851. " -- Now that all of the comments are addressed in this PR (#10861), including your latest comment https://github.com/NVIDIA/spark-rapids/pull/10851#discussion_r1610432557, let's go with this PR. If you don't have any other comments, please approve it ASAP so that I can merge it in?

binmahone commented:
build

binmahone commented:
build

jlowe reviewed (approved):
This looks OK, but should not be merged until either #10851 is merged first or @revans2 and @gerashegalov who had review comments in #10851 are happy with this version of it.

binmahone commented:
> This looks OK, but should not be merged until either #10851 is merged first or @revans2 and @gerashegalov who had review comments in #10851 are happy with this version of it.

The criteria is met (https://github.com/NVIDIA/spark-rapids/pull/10851#issuecomment-2128342806). I'm merging this PR now

binmahone commented:
This PR also closes #10859
